### PR TITLE
adapter: Cancel peeks when clusters are dropped

### LINKED ISSUE #20468: parallel-workload: thread 'coordinator' panicked at 'unable to cancel peek: InstanceMissing(User(41))', /home/deen/git/materialize/src/adapter/src/util.rs:280:23
### What version of Materialize are you using?

90eab10548572f18e6fdeec6e6165ec532a53e6e

### What is the issue?

Using the new WIP parallel-workload with create/drop/set cluster: https://github.com/MaterializeInc/materialize/pull/20354
```
git checkout 90eab10548572f18e6fdeec6e6165ec532a53e6e
bin/environmentd --release --reset
bin/parallel-workload-test --threads=16 --complexity=ddl --scenario=regression --seed=1688671291
```
Can be reproduced in 1 minute on my system:
```
thread 'coordinator' panicked at 'unable to cancel peek: InstanceMissing(User(41))', /home/deen/git/materialize/src/adapter/src/util.rs:280:23
stack backtrace:
   0: rust_begin_unwind
             at /rustc/065852def0903296da33a9eaf557f230bcf3a61a/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/065852def0903296da33a9eaf557f230bcf3a61a/library/core/src/panicking.rs:64:14
   2: <core::result::Result<T,E> as mz_adapter::util::ResultExt<T>>::unwrap_or_terminate
             at ./src/adapter/src/util.rs:280:23
   3: mz_adapter::coord::ddl::<impl mz_adapter::coord::Coordinator>::catalog_transact_with::{{closure}}::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/ddl.rs:378:25
   4: mz_adapter::coord::ddl::<impl mz_adapter::coord::Coordinator>::catalog_transact_with::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/ddl.rs:450:9
   5: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
   6: mz_adapter::coord::ddl::<impl mz_adapter::coord::Coordinator>::catalog_transact_with::{{closure}}
             at ./src/adapter/src/coord/ddl.rs:81:5
   7: mz_adapter::coord::ddl::<impl mz_adapter::coord::Coordinator>::catalog_transact::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/ddl.rs:68:61
   8: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
   9: mz_adapter::coord::ddl::<impl mz_adapter::coord::Coordinator>::catalog_transact::{{closure}}
             at ./src/adapter/src/coord/ddl.rs:62:5
  10: mz_adapter::coord::sequencer::inner::<impl mz_adapter::coord::Coordinator>::sequence_drop_objects::{{closure}}
             at ./src/adapter/src/coord/sequencer/inner.rs:1144:50
  11: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_plan::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/sequencer.rs:202:81
  12: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  13: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_plan::{{closure}}
             at ./src/adapter/src/coord/sequencer.rs:57:5
  14: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute_inner::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:671:72
  15: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  16: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute_inner::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:431:5
  17: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:428:53
  18: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  19: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:375:5
  20: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  21: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_command::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:126:71
  22: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::message_command::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:241:33
  23: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  24: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::message_command::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:238:5
  25: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::handle_message::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:43:63
  26: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  27: mz_adapter::coord::Coordinator::serve::{{closure}}
             at ./src/adapter/src/coord.rs:1524:17
  28: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/park.rs:283:63
  29: tokio::runtime::coop::with_budget
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/coop.rs:107:5
  30: tokio::runtime::coop::budget
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/coop.rs:73:5
  31: tokio::runtime::park::CachedParkThread::block_on
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/park.rs:283:31
  32: tokio::runtime::context::BlockingRegionGuard::block_on
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/context.rs:315:13
  33: tokio::runtime::handle::Handle::block_on
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.27.0/src/runtime/handle.rs:264:9
  34: mz_adapter::coord::serve::{{closure}}::{{closure}}
             at ./src/adapter/src/coord.rs:1752:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### PR DESCRIPTION
**Pull request by jkosh44**
This commit will cancel all pending peek when the cluster running the peek is dropped. When the cluster that is running a peek is dropped, the peek will never complete and will contain a dangling reference to a non-existent cluster.

Fixes #20468

### Motivation
This PR fixes a recognized bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
@def- This fixes my local repro of the issue can you check your test? I'll work on a targeted rust test as well.

**Review by mjibson — APPROVED** [review]

**Review by def- — APPROVED** [review]
Thanks for the fix! Would be great to also add an explicit test for this with the simpler scenario you had. Not sure if SLT is good enough for 2 separate connections.

I'll run parallel-workload against this for a bit, if it falls over with the same panic I'll say so.

**Comment by jkosh44** [comment]
> Thanks for the fix! Would be great to also add an explicit test for this with the simpler scenario you had. Not sure if SLT is good enough for 2 separate connections.
> 
> I'll run parallel-workload against this for a bit, if it falls over with the same panic I'll say so.

Did you post this comment before or after seeing the test I updated? That test is mostly equivalent to the repro I had and will cause a panic on `main`.

File: src/environmentd/tests/sql.rs [hunk 1]
```diff
@@ -2337,24 +2337,28 @@ fn test_coord_startup_blocking() {
 }
 
 #[mz_ore::test]
-fn test_cancel_on_dropped_cluster() {
```
Comment (def-) [inline src/environmentd/tests/sql.rs:2340]: Is it intentional that we don't have the test_cancel_on_dropped_cluster test anymore after this or does the test now cover both regressions?

**Comment by def-** [comment]
> Did you post this comment before or after seeing the test I updated?

Before the test. Great, thanks.

File: src/environmentd/tests/sql.rs (hunk 1 above)
Comment (jkosh44) [reply src/environmentd/tests/sql.rs:2340]: Before this PR, a peek on a dropped cluster would just hang forever and the user would have to manually cancel it. Now when the cluster is dropped all pending peeks are dropped immediately. So it's not possible for a user to cancel the peek on a dropped cluster.

**Pull request merged**